llvm/utils/TableGen/DynamicTableEmitter.cpp

```diff
@@ -0,0 +1,277 @@
+//===- DynamicTableEmitter.cpp - Generate efficiently dynamic tables -==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This tablegen backend emits a generic array initialized by specified fields,
+// together with companion index tables and lookup functions (binary search,
+// currently).
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/TableGen/Error.h"
+#include "llvm/TableGen/Record.h"
+#include "llvm/TableGen/TableGenBackend.h"
+#include <algorithm>
+#include <iostream>
+#include <set>
+#include <sstream>
+#include <string>
+#include <vector>
+
+using namespace llvm;
+
+#define DEBUG_TYPE "dynamic-table-emitter"
+
+namespace {
+
+int getAsInt(Init *B) {
+  return cast<IntInit>(
+             B->convertInitializerTo(IntRecTy::get(B->getRecordKeeper())))
+      ->getValue();
+}
+
+struct GenericField {
+  std::string Name;
+  RecTy *RecType = nullptr;
+  bool IsCode = false;
+  bool IsList = false;
+
+  GenericField(StringRef Name) : Name(std::string(Name)) {}
+};
+
+struct DynamicTable {
+  std::string Name;
+  ArrayRef<SMLoc> Locs; // Source locations from the Record instance.
+  std::string PreprocessorGuard;
+  std::string CppTypeName;
+  SmallVector<GenericField, 2> Fields;
+  std::vector<Record *> Entries;
+
+  const GenericField *getFieldByName(StringRef Name) const {
+    for (const auto &Field : Fields) {
+      if (Name == Field.Name)
+        return &Field;
+    }
+    return nullptr;
+  }
+};
+
+class DynamicTableEmitter {
+  RecordKeeper &Records;
+  std::set<std::string> PreprocessorGuards;
+
+public:
+  DynamicTableEmitter(RecordKeeper &R) : Records(R) {}
+
+  void run(raw_ostream &OS);
+
+private:
+  std::string primaryRepresentation(SMLoc Loc, const GenericField &Field,
+                                    Init *I) {
+    if (StringInit *SI = dyn_cast<StringInit>(I)) {
+      if (Field.IsCode || SI->hasCodeFormat())
+        return std::string(SI->getValue());
+      else
```

Comment:
That's funny because this is actually copy-pasted from https://github.com/intel/llvm/blob/c87e7802ef8087e81d66c41819fa05ab76fbd408/llvm/utils/TableGen/SearchableTableEmitter.cpp#L117. I'll change it in the new file.